### ISSUE TITLE
Detaching l_t

### ISSUE THREAD

**Issue by Pozimek** [kevinzakka/recurrent-visual-attention#29]
At the moment the location tensor _l_t_ is never detached from the computational graph in spite of both being produced by and 'consumed' by trainable modules. As far as I understand the code this enables the gradients to 'backpropagate through time' in a way that the authors of RAM did not intend: the gradients that originated in the _action_network_ and reached the _fc2_ layer inside the _glimpse network_ would travel back to the previous timestep's _location_network_ and alter its weights and only stop once they reach the detached RNN memory vector _h_t_. As far as I understand the authors intended the _location_network_ to only be trained using reinforcement learning.

This could be a bug or it could be an accidental improvement to the network; either way please let me know if my understanding is correct in here as I am still learning Pytorch and my project is heavily reliant on your code :)

**Comment by yxiao54** [comment]
Yes agree. Same confusion. The author says:**The location network  is always trained with REINFORCE.** So should we build another loss function?

**Comment by lijiangguo** [comment]
Note aside from stop at h_t, the gradient originated from action_network will continue recursively through g_t in core_network to modify all previous time l_t. Meanwhile, I wonder why location_network and baseline_network have to detach from h_t? Anywhere in the paper suggested core_network is only trained via classification loss? @Pozimek @yxiao54

**Comment by litingfeng** [comment]
@Pozimek it seems that l_t is detached in [location network](https://github.com/kevinzakka/recurrent-visual-attention/blob/a38ac8958ebf1c61a10c4d5320f1e31d3d0b73dd/modules.py#L313)

**Comment by lizhenstat** [comment]
@Pozimek Hi, you explanation helps me understand why the authors use l_t.detach() in the code, thanks!
